Poc/react rsc (#508)
storyblok/storyblok-react: README.md

	</a>
	<h1 align="center">@storyblok/react</h1>
  <p align="center">

>     The React plugin you need to interact with <a href="http://www.storyblok.com?utm_source=github.com&utm_medium=readme&utm_campaign=storyblok-react" target="_blank">Storyblok API</a> and enable the <a href="https://www.storyblok.com/docs/guide/essentials/visual-editor?utm_source=github.com&utm_medium=readme&utm_campaign=storyblok-react" target="_blank">Real-time Visual Editing Experience</a>. This package helps you integrate Storyblok with React along with all types of Next.js projects. 

Author: I think we can also mention Remix. Do you think it would be good to mention Gastby as we already have a separate SDL for that?

>     The React plugin you need to interact with <a href="http://www.storyblok.com?utm_source=github.com&utm_medium=readme&utm_campaign=storyblok-react" target="_blank">Storyblok API</a> and enable the <a href="https://www.storyblok.com/docs/guide/essentials/visual-editor?utm_source=github.com&utm_medium=readme&utm_campaign=storyblok-react" target="_blank">Real-time Visual Editing Experience</a>. This package helps you integrate Storyblok with React along with all types of Next.js projects. 

Reviewer: Yes, that makes sense. My main point is that the SDK is not "limited to Next.js", but it can be used in any React project

  components: {
    page: Page,
    teaser: Teaser,
  },
});
```

> > Note: This is the general way for initalizing the SDK, the initialization might be different depending upon the framework. You can see how everything works according to the framework in their respective sections below.

Reviewer: As we're only showing the `storyblokInit` call here, I think it will be pretty much the same for all the frameworks. What it would change is where/when to call it, right?

Author: In the last case, where we enable the live editing we don't add the components and initialize it twice.

Reviewer: 👍

> Note: For spaces created in the United States or China, the `region` parameter **must** be specified.

`@storyblok/react` does three actions when you initialize it:

- Provides a `getStoryblokApi` object in your app, which is an instance of [storyblok-js-client](https://github.com/storyblok/storyblok-js-client).

> - Loads [Storyblok Bridge](https://www.storyblok.com/docs/Guides/storyblok-latest-js?utm_source=github.com&utm_medium=readme&utm_campaign=storyblok-react) for real-time visual updates.

Reviewer: ```suggestion
- Loads the [Storyblok Bridge](https://www.storyblok.com/docs/Guides/storyblok-latest-js?utm_source=github.com&utm_medium=readme&utm_campaign=storyblok-react) for real-time visual updates.
```

  return (
    <div {...storyblokEditable(blok)} key={blok._uid} data-test="feature">
      <div>
        <div>{blok.name}</div>
        <p>{blok.description}</p>
      </div>
    </div>
  );
};

export default Feature;
```

Where `blok` is the actual blok data coming from [Storyblok's Content Delivery API](https://www.storyblok.com/docs/api/content-delivery?utm_source=github.com&utm_medium=readme&utm_campaign=storyblok-react).

> Note: The `storyblokEditable` function works the same way for all the frameworks and components created.

### Getting Started

**This SDK provides you the support to work with React, Next.js 12 (or Next.js 13 with no RSC), Next.js 13 (Full RSC) and Next.js (Partial RSC, with Live Editing). Depending upon these different frameworks and versions, the way to use the SDK and the functionalities it provides differ.**

Below is the guide on how to use it with all the frameworks -

## React

The initalization remains the same when you work with React. You can intialze the SDK in the `index.js` file. Please refer to the 'Initialization' section above to read more.

### Fetching Content and Listening to Storyblok Visual Editor events

> For React, use `useStoryblok` to fetch the content as well as enable the live editing. You need to pass the `slug` as first parameter, and `apiOptions` as second parameter and `bridgeOptions` as third parameter which is optional if you want to set the options for bridge by yourself:

Reviewer: ```suggestion
Use `useStoryblok` to fetch the content as well as enable live editing. You need to pass the `slug` as the first parameter, `apiOptions` as the second parameter, and `bridgeOptions` as the third parameter, which is optional if you want to set the options for the bridge by yourself:
```

Reviewer: also, should we explain what `apiOptions` and `bridgeOptions` mean and represent?

Author: Do you think this would be enough - Check the available [apiOptions](https://github.com/storyblok/storyblok-js-client#class-storyblok) (passed to `storyblok-js-client`) and [bridgeOptions](https://www.storyblok.com/docs/Guides/storyblok-latest-js?utm_source=github.com&utm_medium=readme&utm_campaign=storyblok-nuxt) (passed to the Storyblok Bridge). ?

Reviewer: 👌

export default App;
```

`StoryblokComponent` renders the route components dynamically, using the list of components loaded during the initialization inside the `storyblokInit` function. 

This is how you can pass the Bridge options as a third parameter to `useStoryblok`:

```js
useStoryblok(story.id, {version: 'draft'}, {
  resolveRelations: ["Article.author"],
});
```

**Check out our React Boilerplate [here](https://github.com/storyblok/storyblok-react-boilerplate), or read on how to add Storyblok to React in 5 mins [here](https://www.storyblok.com/tp/headless-cms-react)**
You can also take a look at the [React Playground](https://github.com/arorachakit/storyblok-react/tree/main/playground) in this repo.

> ## Next.js using App Router - Full React Server Components

Reviewer: @arorachakit as discussed on Slack, should we swap this section with "Next.js using App Router - Live Editing support"?

Author: @fgiuliani I've swap the sections and adjusted the content according to it. Please take a look.

export default App;
```

> `StoryblokComponent` here renders all the components dynamically which you loaded before during the initalization inside the `storyblokInit` function. 
> 
> This is how you can pass the Bridge options as a third parameter to `useStoryblok` :

Reviewer: ```suggestion
`StoryblokComponent` renders the route components dynamically, using the list of components loaded during the initialization inside the `storyblokInit` function. 

This is how you can pass the Bridge options as a third parameter to `useStoryblok` :
```

```js
useStoryblok(story.id, {version: 'draft'}, {
  resolveRelations: ["Article.author"],
});
```

> **Check out the React Bolierplate [here](https://github.com/storyblok/storyblok-react-boilerplate) or read on how to add Storyblok to React in 5 mins [here](https://www.storyblok.com/tp/headless-cms-react)**

Reviewer: ```suggestion
**Check out our React Boilerplate [here](https://github.com/storyblok/storyblok-react-boilerplate), or read on how to add Storyblok to React in 5 mins [here](https://www.storyblok.com/tp/headless-cms-react)**
```

> ## Next 12 and Next 13 (no RSC)

Reviewer: Actually, this would be pretty much the same approach to take for a Remix project. Should we add "Remix" to the title? Or create a separate section for it?

> ## Next 12 and Next 13 (no RSC)

Reviewer: Thinking out loud: As the App Router is stable now, and probably the "official approach", should we move this section below on the page so the devs see first the App Router approach?

> ## Next 12 and Next 13 (no RSC)

Author: Yes! I agree to all these points! 
Also, let's add remix!

> ## Next 12 and Next 13 (no RSC)

Author: I've made the change, please take a look.

> ## Next 12 and Next 13 (no RSC)

Author: I haven't checked if I need to add something specific for Remix.

> ## Next 12 and Next 13 (no RSC)

Reviewer: Just to summarize: I discussed with @arorachakit and we will add Remix example to the README file in a separate PR, after we release the RSC support

> In this section, the way to use the SDK with the `pages` directory is described.

Reviewer: ```suggestion
In this section, we'll see how to use the React SDK with the `pages` directory approach.
```

The initalization remains the same when you work with Next.js. You can intialze the SDK in the `_app.js` file. Please refer to the 'Initialization' section above to read more.

### Fetching Content

> The SDK provdies a `getStoryblokApi` object in your app, which is an instance of [storyblok-js-client](https://github.com/storyblok/storyblok-js-client). This can be used to fetch the content from Storyblok. You can use it in functions like `getStaticProps`, `getStaticPaths`, `getServerSideProps` etc.

Reviewer: ```suggestion
The SDK provides a `getStoryblokApi` object in your app, which is an instance of [storyblok-js-client](https://github.com/storyblok/storyblok-js-client). This can be used to fetch the content from Storyblok. You can use it in functions like `getStaticProps`, `getStaticPaths`, `getServerSideProps` etc.
```

```js
import { getStoryblokApi } from "@storyblok/react";

// At the required place
const storyblokApi = getStoryblokApi();
const { data } = await storyblokApi.get("cdn/stories", { version: "draft" });
```

> > Note: To use this, you need to use the `apiPlugin` during the initalization inside the `storyblokInit` function. If you don't use `apiPlugin`, you can use your prefered method or function to fetch your data.

Reviewer: ```suggestion
> Note: To use this approach you need to include the `apiPlugin` module when calling the `storyblokInit` function. If you don't use `apiPlugin`, you can use your preferred method or function to fetch your data.
```

### Listening to Storyblok Visual Editor events

> The SDK also gives you `useStoryblokState` hook. It works similar to `useStoryblok` for live editing, but doesn't fetch the content. Instead, it takes a story object as the first parameter. You can also pass the Bridge Options as the second paramter.

Reviewer: ```suggestion
The SDK also provides you with the `useStoryblokState` hook. It works similarly to `useStoryblok` for live editing but it doesn't fetch the content. Instead, it receives a story object as the first parameter. You can also pass the Bridge Options as the second parameter.
```

```js
import {
  useStoryblokState,
  StoryblokComponent,
} from "@storyblok/react";

export default function Home({ story: initialStory }) {
  const story = useStoryblokState(initialStory);

  if (!story.content) {
    return <div>Loading...</div>;
  }

  return <StoryblokComponent blok={story.content} />;
}
```

> In this case, it story is being passed as a prop that can be coming from where the story is being fetched. A complete example would look like this- 

Reviewer: ```suggestion
In this case, the story is being passed as a prop that can be coming from where the story is being fetched. A complete example would look like this:
```

}
```

> `StoryblokComponent` renders all the components dynamically which you loaded before during the initalization inside the `storyblokInit` function. 

Reviewer: ```suggestion
`StoryblokComponent` renders the route components dynamically, using the list of components loaded during the initialization inside the `storyblokInit` function. 
```

> **Check out the this [repo](https://github.com/storyblok/next.js-ultimate-tutorial/tree/part-1), it is the code for the first part of our Ultimate Tutorial for Next.js or read on how to add Storyblok to Next.js in 5 mins [here](https://www.storyblok.com/tp/add-a-headless-cms-to-next-js-in-5-minutes)**

Reviewer: ```suggestion
**Check out the [code for the first part of our Next.js + Storyblok Ultimate Tutorial](https://github.com/storyblok/next.js-ultimate-tutorial/tree/part-1). Or you can also read on how to add Storyblok to a Next.js project in 5 minutes [here](https://www.storyblok.com/tp/add-a-headless-cms-to-next-js-in-5-minutes)**
```

> ## Next 13 (Full RSC)

Reviewer: ```suggestion
## Next.js using App Router - Full React Server Components
```

> If you're using the complete React Server Components along with app directory, follow this method. Using all the Server Components has great performance benefits. 

Reviewer: ```suggestion
If you're using the Next.js `app` directory approach, and React Server Components exclusively, follow this approach.
```

> > The SDK has a special module for RSC. Always use `@storyblok/react/rsc` for importing while using Server Side Components.

Reviewer: ```suggestion
> The SDK has a special module for RSC. Always import `@storyblok/react/rsc` while using Server Components.
```

> **Limitation** - Real time editing can not work if all the components are on the server, you can see the changes when you save the story.

Reviewer: ```suggestion
**Limitation** - Real-time editing won't work if all the components are rendered on the server. Although, you can see the changes applied in the Visual Editor whenever you save or publish the story.
```

### Initialize and Import your Storyblok Components

The initialzation remains the same here as well. Please refer to the above section about "Initialization" for more information about `storyblokInit` function.

> In `_app/layout.jsx`, call `storyblokInit` and use the new `StoryblokBridgeLoader` to setup the bridge. This Bridge Loader can be imported from `@storyblok/react/bridge-loader`:

Reviewer: ```suggestion
In `_app/layout.jsx`, call the `storyblokInit` function and use the new `StoryblokBridgeLoader` component to setup the Storyblok bridge. This Bridge Loader can be imported from `@storyblok/react/bridge-loader`:
```

```js
import { storyblokInit, apiPlugin } from "@storyblok/react/rsc";
import StoryblokBridgeLoader from "@storyblok/react/bridge-loader";

import Page from "../components/Page";
import Teaser from "../components/Teaser";

storyblokInit({

>   accessToken: "RVLbQ7MG4kCi3r1sQQ4VEAtt",

Reviewer: ```suggestion
  accessToken: "YOUR_ACCESS_TOKEN",
```

  use: [apiPlugin],
  components: {
    teaser: Teaser,
    page: Page,
  },
});

export default RootLayout({ children }) =>{
  return (
    <html lang="en">
      <body>{children}</body>
      <StoryblokBridgeLoader options={bridgeOptions} />
    </html>
  );
}
```
As the name says, `StoryblokBridgeLoader` loads the bridge on the client. It helps you see the dotted lines and allows you to still click on the components inside the Visual Editor to open their schema. You can pass the bridge options using the `options` prop.

### 2. Fetch Content and Render Components

> The `getStoryblokApi` can be used to fetch the data from Storyblok. This is imported from `@storyblok/react/rsc`.

Reviewer: ```suggestion
The `getStoryblokApi` function can be used to fetch the data from the Storyblok API. This is imported from `@storyblok/react/rsc`.
```

> In `_app/page.jsx`, use it as follows:

Reviewer: ```suggestion
Go to the route you want to fetch data from and use it as follows:
```

```js
import { getStoryblokApi, StoryblokComponent } from "@storyblok/react/rsc";

export default async function Home() {
  const { data } = await fetchData();

  return (
    <div>
      <h1>Story: {data.story.id}</h1>
      <StoryblokComponent blok={data.story.content} />
    </div>
  );
}

export async function fetchData() {
  const storyblokApi = getStoryblokApi();
  return storyblokApi.get(`cdn/stories/home`, { version: "draft" });
}
```

> `StoryblokComponent` here renders all the components dynamically which you loaded before during the initalization inside the `storyblokInit` function. 

Reviewer: ```suggestion
`StoryblokComponent` renders the route components dynamically, using the list of components loaded during the initialization inside the `storyblokInit` function.
```

**Demo: see _playground-next13-rsc_**

> ## Next 13 (with Live Editing)

Reviewer: ```suggestion
## Next.js using App Router - Live Editing support
```

> You can enable the live editing with Server Side Components by rendering them with a wrapper (`StoryblokPovider`) on client. The SDK allows you to take full advantage of the Live Editing, but the use of Server Side Components is partial, which is still better than Next.js 12 performance wise. 

Reviewer: ```suggestion
You can enable Storyblok Visual Editor's live editing with React Server Components by rendering them inside a wrapper (`StoryblokPovider`) on the client. The SDK allows you to take full advantage of the Live Editing, but the use of Server Side Components is partial, which will be still better than the older Next.js approach performance-wise. 
```

> > The SDK has a special module for RSC. Always use `@storyblok/react/rsc` for importing while using Server Side Components.

Reviewer: ```suggestion
> The SDK has a special module for RSC. Always import `@storyblok/react/rsc` while using Server Components.
```

### 1. Initialize

> In `_app/layout.jsx`, call `storyblokInit` without loading the components and use `StoryblokProvider` (this provider is created in the next step) :

Reviewer: ```suggestion
In `app/layout.jsx`, call the `storyblokInit` function, but without loading the component list (we will do that on the client). Wrap your whole app using a `StoryblokProvider` component (this provider is created in the next step) :
```

```js
import { storyblokInit, apiPlugin } from "@storyblok/react/rsc";
import StoryblokProvider from "../components/StoryblokProvider";

storyblokInit({

>   accessToken: "OurklwV5XsDJTIE1NJaD2wtt",

Reviewer: ```suggestion
  accessToken: "YOUR_ACCESS_TOKEN",
```

  use: [apiPlugin],
});

export default function RootLayout({ children }) {
  return (
    <StoryblokProvider>
      <html lang="en">
        <body>{children}</body>
      </html>
    </StoryblokProvider>
  );
}
```

### 2. Create StoryblokProvider and Import your Storyblok Components

> Create `components/StoryblokProvider.jsx`, and re-initalize and import your Storyblok components:

Reviewer: ```suggestion
Create the `components/StoryblokProvider.jsx` file. Re-initalize the connection with Storyblok (this time, on the client) using `storyblokInit`, and import your Storyblok components:
```

```js

> /** 1. Tag it as client component */

Reviewer: ```suggestion
/** 1. Tag it as a client component */
```

"use client";
import { storyblokInit, apiPlugin } from "@storyblok/react/rsc";

/** 2. Import your components */
import Page from "../components/Page";
import Teaser from "../components/Teaser";

/** 3. Initialize it as usual */
storyblokInit({

>   accessToken: "OurklwV5XsDJTIE1NJaD2wtt",

Reviewer: ```suggestion
  accessToken: "YOUR_ACCESS_TOKEN",
```

  use: [apiPlugin],
  components: {
    teaser: Teaser,
    page: Page,
  },
});

export default function StoryblokProvider({ children }) {
  return children;
}
```

> > Note: it's necessary to re-initialize here as well, as to enable the live editing you must initialize the lib universally (client + server) via a React client component.

Reviewer: ```suggestion
> Note: it's necessary to re-initialize here as well, as to enable the live editing experience inside the Visual Editor you need to initialize the lib universally (client + server).
```

### 3. Fetch Content and Render Components

> The `getStoryblokApi` can be used to fetch the data from Storyblok. This is imported from `@storyblok/react/rsc`. Instead of using `StoryblokCompnent`, use `StoryblokStory` which will automatically handle the live event of the story. In `_app/page.jsx`, use them as follows:

Reviewer: ```suggestion
The `getStoryblokApi` can be used to fetch the data from the Storyblok API. This should be imported from `@storyblok/react/rsc`. 
Instead of using `StoryblokCompnent`, you can render the content of your route with the `StoryblokStory` component, which will automatically handle the Visual Editor live events when editing the story. In `app/page.jsx`, use them as follows:
```

```js
import { getStoryblokApi } from "@storyblok/react/rsc";
import StoryblokStory from "@storyblok/react/story";

export default async function Home() {
  const { data } = await fetchData();

  return (
    <div>

>       <h1>Story: {data.story.id}</h1>

Reviewer: I think we can remove this line

      <StoryblokStory story={data.story} />
    </div>
  );
}

export async function fetchData() {
  const storyblokApi = getStoryblokApi();
  return storyblokApi.get(`cdn/stories/home`, { version: "draft" });
}
```
**Demo: see _playground-next13-live-editing_**

> ### Features and API

Reviewer: ```suggestion
## Features and API
```

You can **choose the features to use** when you initialize the plugin. In that way, you can improve Web Performance by optimizing your page load and save some bytes.

> #### Storyblok API

Reviewer: ```suggestion
### Storyblok API
```

Reviewer: We should do the same with `#### Storyblok Bridge`

sbBridge.on(["input", "published", "change"], (event) => {
  // ...
});

> ```

Reviewer: This is wrong, it's wrapping the following text section into a code block

> #### Rendering Rich Text

Reviewer: ```suggestion
### Rendering Rich Text
```
